File: 166. Fraction to Recurring Decimal.py
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        a, b = numerator, denominator
        flag = ''
        if a * b < 0:
            flag = '-'
        a, b = abs(a), abs(b)
        c = a // b
        d = a % b
        if d == 0:
            return flag + str(c)
        x = str(c)
        y = ''
        ds = set()
        dl = []
        while d != 0:
            if d in ds:
                i = dl.index(d)
                ret = flag + x + '.' + y[:i] + '(' + y[i:] + ')'
                return ret
            dl.append(d)
            ds.add(d)
            a = d * 10
            c = a // b
            d = a % b
            y += str(c)
            if d == 0:
                return flag + x + '.' + y

File: 166. Fraction to Recurring Decimal.go
func fractionToDecimal(numerator int, denominator int) string {
	var num, deno int64 = int64(numerator), int64(denominator)
	if num%deno == 0 {
		return strconv.FormatInt(num/deno, 10)
	}

	var res string
	if (num < 0) != (deno < 0) {
		res += "-"
	}

	if num < 0 {
		num = -num
	}

	if deno < 0 {
		deno = -deno
	}

	res += strconv.FormatInt(num/deno, 10)
	res += "."

	num = num % deno
	nums := make(map[int64]int)
	for num != 0 {
		if _, ok := nums[num]; ok {
			res = res[:nums[num]] + "(" + res[nums[num]:] + ")"
			break
		}

		nums[num] = len(res)
		num *= 10
		res += strconv.FormatInt(num/deno, 10)
		num %= deno
	}
	return res
}